models/flowers_oop.nim

```diff
@@ -0,0 +1,510 @@
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+uniform samplerCube environmentMap;
+in vec3 fragPosition;
+in vec3 t;
+in vec3 tr;
+in vec3 tg;
+in vec3 tb;
+in float rfac;
+
+out vec4 finalColor;
+
+void main() {
+
+    vec4 ref = textureCube(environmentMap, t);
+    vec4 ret = vec4(1.0);
+
+    ret.r = textureCube(environmentMap, tr).r;
+    ret.g = textureCube(environmentMap, tg).g;
+    ret.b = textureCube(environmentMap, tb).b;
+    finalColor = ret * rfac + ref * (1.0 - rfac);
+}
+"""
+
+const
+  MaxFlowerParam = 20.0
+  MinFlowerParam = -20.0
+  VelocityStep = 0.02
+  VelocityBias = 0.01
+
+type Flower = object
+  cubemap: TextureCubemap
+  chromaticDispersion: Vector3
+  bias: float32
+  scale: float32
+  power: float32
+  a: float32
+  b: float32
+  c: float32
+  d: float32
+  tdelta: float32
+  pdelta: float32
+  cubemapModelLoc: ShaderLocation
+  chromaticDispersionLoc: ShaderLocation
+  biasLoc: ShaderLocation
+  scaleLoc: ShaderLocation
+  powerLoc: ShaderLocation
+  aLoc: ShaderLocation
+  bLoc: ShaderLocation
+  cLoc: ShaderLocation
+  dLoc: ShaderLocation
+  tdeltaLoc: ShaderLocation
+  pdeltaLoc: ShaderLocation
+  viewLoc: ShaderLocation
+  model: Model
+  modelShader: Shader
+  translation: Vector3
+  velocity: Vector3
+
+proc `=destroy`(self: Flower) =
+  # Clean up resources when object is destroyed
+  if self.cubemap.id != 0:
+    unloadTexture(self.cubemap.id)
+  #if self.modelShader.id != 0:
+  #  unloadShader(self.modelShader.id)
+  #if self.model.meshCount != 0:
+  #  unloadModel(self.model.id)
+
+proc `=copy`(dest: var Flower, source: Flower) {.error.} =
+  # Disable copying since we manage unique resources
+  discard
+
+proc `=sink`(dest: var Flower, source: Flower) =
+  # Move resources from source to destination
+  if dest.cubemap.id != 0:
+    unloadTexture(dest.cubemap.id)
+  #if dest.modelShader.id != 0:
+  #  unloadShader(dest.modelShader.id)
+  #if dest.model.meshCount != 0:
+  #  unloadModel(dest.model.id)
+
+  dest = source
+  # Zero out source to prevent double-free
+  #source.cubemap.id = 0
+  #source.modelShader.id = 0
+  #source.model.meshCount = 0
```

Comment:
All these destructor hooks should be removed. The raylib library handles destructors for you. (If you need to attach a destructor to an object that has other managed resources you call them explicitly on the fields).